Migration of  hardhat-upgrade to plugin to use hardhat v3

afa7789 opened the pull request:
⚠️ Changeset and versioning must be reviewed so we don't break versioning for users of previous version too ( hh2). ⚠️ 

# Migration to Hardhat 3: ESM & Plugin Hooks

This PR migrates `@openzeppelin/hardhat-upgrades` to **Hardhat 3**, converting the package to **ESM** and adopting the new **plugin hooks architecture**. This is a major breaking change that replaces the legacy `extendEnvironment` pattern with `HardhatPlugin` hook handlers for config and Solidity initialization. Users must now utilize explicit async factory functions—`await upgrades(hre, connection)` and `defender()`—and manage network connectivity via `await hre.network.connect()`, as `ethers` is now obtained through the connection rather than the HRE.

- Technical Implementation & Compatibility

The package has been fully converted to ESM (`type: module`) using `Node16` resolution. To ensure compatibility with Babel and NYC coverage, `import...with { type: 'json' }` statements were replaced with `createRequire()` across 15 files. Dependencies are updated to `hardhat@^3.0.16`, `ethers v6`, and the latest `@nomicfoundation` plugins.

Testing infrastructure now includes Solidity test support via the `foundry-upgrades` repository. To stabilize the environment, example projects were isolated using `pnpm` link protocols and local AVA configurations. CI/CD workflows were modified to initialize git submodules before installation and now include `foundry-upgrades` in `npmFilesToBuild`.

- Documentation & Known Constraints

A comprehensive `MIGRATION.md` has been added with before/after examples, and the README now reflects Hardhat 3 patterns. Please note that full `hardhat-verify` integration and certain namespace layouts are currently pending an upstream fix in Hardhat ([[PR #7744](https://github.com/NomicFoundation/hardhat/pull/7744)](https://github.com/NomicFoundation/hardhat/pull/7744)), which also impacts the provided example projects. While core functionality is tested and passing, some edge cases in verification may require further refinement following the upstream merge. We also need a new feature from the nomic team to allow access to their internal libraries for us to use them like in hh2, at the verify part.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  - Added Hardhat 3 support with ESM module structure and plugin hooks
  - Updated tooling to Node 22 and ethers v6.14.0
  - Introduced example projects for Transparent and UUPS proxy patterns

* **Documentation**
  - Added migration guide for Hardhat 2 to 3 upgrade
  - Updated README with TypeScript-first approach and connection-sharing patterns
  - Documented Hardhat 3 API changes and configuration updates

* **Chores**
  - Updated build configuration and dependencies for Hardhat 3 compatibility
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

afa7789 committed:
chore(plugin): initial changes for hardhat 3

packages changes and initial config in accord to migration documentation.

afa7789 committed:
chore(plugin): second change on tsconfig.json to be able to avoid tserrors for ESM structure.

afa7789 committed:
chore(dev):  upgrading files to use hardhat3.

importing errors + a few todos, change of Imports + new functions  to use hardhat3 library in plugin files.

afa7789 committed:
chore: lint fix

afa7789 committed:
chore(plugin): update imports to use .js extensions and adjust tsconfig for NodeNext module resolution

ericglau commented on packages/plugin-hardhat/src/index.ts:
```diff
@@ -1,29 +1,37 @@
 /* eslint-disable @typescript-eslint/no-var-requires */
+// TODO figure out what is the newer, TASK_COMPILE_SOLIDITY and TASK_COMPILE_SOLIDITY_COMPILE
```
These are Hardhat internal subtasks, where Hardhat calls TASK_COMPILE_SOLIDITY -> TASK_COMPILE_SOLIDITY_COMPILE_JOBS -> TASK_COMPILE_SOLIDITY_COMPILE_JOB ->  TASK_COMPILE_SOLIDITY_COMPILE.  And only certain parts of those need to be overridden in our plugin.  See https://github.com/NomicFoundation/hardhat/blob/09951afd78e49465f7949478919c9217a12e8338/packages/hardhat-core/src/builtin-tasks/compile.ts#L1366 for the v2 internal subtasks.

We may need to find the equivalents in v3.

afa7789 committed:
chore(plugin): refactor plugin structure with hooks and update type extensions for Hardhat compatibility

( still have to test it )

afa7789 committed:
chore(plugin): refactor hooks to improve user config resolution and enhance artifact cleanup process

afa7789 committed:
chore(plugin): enhance validation handling and namespaced storage layout support in Solidity hooks

afa7789 committed:
chore(plugin): update dependencies handling and enhance type extensions for defender integration

afa7789 committed:
chore(plugin): lint fixes.

afa7789 committed:
chore(plugin): update import statements to use ES module syntax for JSON files

afa7789 committed:
chore(plugin): update Hardhat configuration and fix plugin integration with upgrades

afa7789 committed:
chore(logs): debugging

afa7789 committed:
chore(provider): adding one more log which should help.

afa7789 committed:
chore(call): enhance logging for callOptionalSignature and add implementation address tests

afa7789 committed:
chore(test): migrate to ES module syntax and update upgrades import

afa7789 committed:
feat: refactor deploy and upgrade functions to accept NetworkConnection

- Updated deploy and upgrade functions across multiple files to accept a NetworkConnection parameter.

afa7789 committed:
refactor(tests): migrate from CommonJS to ES modules in Hardhat tests

afa7789 committed:
chore(tests): update imports and initialization in Hardhat tests

afa7789 committed:
chore: clean up TODO comments and remove unnecessary code in deploy, solidity, etherscan-api, and verify-proxy files

afa7789 committed:
feat: add warning for locked validations cache during compilation

afa7789 committed:
refactor(tests): replace proxyquire with esmock for module mocking

afa7789 committed:
refactor(tests): replace network provider calls with ethers.provider for better context management

afa7789 committed:
refactor(tests): replace hre.ethers with ethers for consistency in contract interactions. Also other fixes needed to make sure these would work.

afa7789 committed:
refactor(tests): migrate mockDeploy function to a new file and update imports across test files

afa7789 committed:
feat(tests): enhance test script to separate Solidity and JavaScript tests, and add new test file for Upgrades library

afa7789 committed:
feat(tests): add WithConstructor and WithConstructorArray contracts, enhance test script for better argument handling

afa7789 committed:
refactor(tests): improve build info file handling and update test script for better artifact management

afa7789 committed:
refactor(tests): enhance logging in Upgrades test and improve argument handling in test script

afa7789 committed:
feat(tests): add new contracts for testing and update hardhat configuration with dotenv support

afa7789 committed:
feat(tests): update Solidity contracts for testing and inject AST into artifacts for Hardhat 3 compatibility

afa7789 committed:
feat(tests): add GreeterInitializable contract and update tests for compatibility

afa7789 committed:
chore(tests):  the changes regarding to solidity ended up breaking the hh tests, needing this commit to fix this.

afa7789 committed:
docs: add migration guide for Hardhat 2 to Hardhat 3 with breaking changes and usage examples

afa7789 committed:
chore(contracts): Refactor and add new contracts for upgradeable patterns

I moved the contracts to a more suitable place for better usability. But this will force us to do at least a small change at the foundry repo.

afa7789 committed:
feat(contracts): add GreeterInitializable contract with initializer function
feat(tests): update tests to use GreeterInitializable for UUPS and Transparent proxies
chore: update submodule for openzeppelin-foundry-upgrades

afa7789 committed:
feat: add iginition modules and deployment.

afa7789 committed:
chore(ux): ux improve on checking tests.

afa7789 committed:
refactor: remove ignition cause it doesn't suit the plugin in itself.

afa7789 committed:
feat: implement OZfilesToBuild utility for OpenZeppelin contract paths

afa7789 committed:
feat: update upgrade paths in Greeter contracts and tests for consistency

afa7789 committed:
feat: update node version to 22.x in GitHub action setup

afa7789 committed:
chore: yarn lint fix

afa7789 committed:
chore(lint): fixed a bunch of lint issues.

afa7789 committed:
chore(lint): more lint fixes

afa7789 committed:
fix: update import paths and reorder dotenv configuration in hardhat config, to fix module import error.

afa7789 committed:
chore(fix): fixing tests that were broken due to contract name changes, etc.

afa7789 committed:
chore(test): make test of example at least run correctly, so I have set the enviroment for it.

But this is not finished yet, as the tests doesn't work, it's related to storage layout.

afa7789 committed:
chore(ci/cd): coverage and lint fixes.

afa7789 committed:
fix: add missing OpenZeppelin foundry upgrade contract paths to npmFilesToBuild

afa7789 committed:
fix: ensure submodules are checked out recursively in workflow steps

afa7789 committed:
chore: update version to 4.0.0 and enhance changelog for Hardhat 3 migration ( I think we need REVIEW )

afa7789 committed:
chore(ci/cd): to avoid crash, this may need to be reverted in CI/CD after we update package.json to not use submodules and yes the correct package version ( waiting our release ).

afa7789 committed:
Updating hardhat to new version with hook requested!

afa7789 committed:
refactor: simplify return statement in callOptionalSignature function

afa7789 committed:
chore: lint fix

afa7789 committed:
chore: update dependencies for foundry-upgrades and upgrades-core

afa7789 committed:
chore: update foundry-upgrades dependency to use local submodule path and bump version to 0.5.0-alpha.0

afa7789 committed:
chore: remove TODO after review.

afa7789 committed:
chore: update foundry-upgrades dependency version to 0.4.1-alpha.0 and adjust yarn.lock

afa7789 committed:
chore: update Node.js version to 22 and downgrade @openzeppelin/upgrades-core to 1.45.0-alpha.0

afa7789 committed:
chore: bump @openzeppelin/upgrades-core version to 1.45.0-alpha.0 and clean up yarn.lock

ericglau commented:
@coderabbitai review

afa7789 committed:
chore: tests connection clean up.

afa7789 committed:
chore: update migration guide for Hardhat 3 and enhance test script error handling

afa7789 committed:
chore: reorder imports for consistency across test files

afa7789 committed:
chore: add strict mode and source map options to TypeScript configuration

afa7789 committed:
chore: update TokenV2 to retain OwnableUpgradeable for storage compatibility and adjust migration function

afa7789 committed:
chore: enhance validation for build-info files in Foundry and Hardhat 3 formats

afa7789 committed:
chore: testing CI/CD "before releasing"

afa7789 committed:
chore: update comments in checks.yml for clarity on core package build process

afa7789 committed:
chore: fix symlink for local core CLI in plugin-hardhat to resolve binary conflicts

afa7789 committed:
refactor: remove Etherscan API config mock and add verification mock for Hardhat 3

- Removed the mock for Etherscan API configuration in the test setup.
- Introduced a mock for verification.etherscan to support Hardhat 3 (hardhat-verify v3.0.10+).

afa7789 committed:
feat: add force-local-core-cli action and update checks to use it

afa7789 committed:
refactor: simplify test script by removing redundant argument checks

afa7789 committed:
fix: cast connection to unknown type for Etherscan verification access

afa7789 committed:
feat: add caching for Hardhat solc compilers in setup action

ericglau committed:
Improve build info fallback handling for Hardhat 2

ericglau committed:
Bump alpha version for core

ericglau committed:
Set alpha version for hardhat plugin, update changelog to be user facing

ericglau committed:
Update peer deps, update readme and migration

ericglau committed:
Simplify example projects

ericglau committed:
fix dependency conflicts

ericglau committed:
Update examples

ericglau committed:
Update readme for next tag and clarify ethers/viem

ericglau committed:
Update packages to real versions to be published

ericglau committed:
rename example to examples

ericglau committed:
Move verify into subplugin, remove from peerDependencies

ericglau committed:
Update changelog and package for verify

ericglau committed:
update readme and changelog for peer dependencies handling

ericglau committed:
Use semver for hardhat devDep

ericglau committed:
Improve examples

ericglau committed:
Add test script for examples

ericglau committed:
Fix etherscan verification to use latest hardhat-verify API

ericglau committed:
Fix etherscan custom API calls

ericglau committed:
Update migration guide for verify

ericglau committed:
Use consistent upgradesApi in examples

ericglau committed:
Add note in top readme

ericglau committed:
Add notes for alpha status

ericglau committed:
Remove unused package scripts

ericglau commented:
@SocketSecurity ignore-all
These look like low-severity heuristic alerts on common dependencies (mostly dev/test tooling).

ericglau committed:
Test logging upgrades core resolution inatead of forcing local

ericglau committed:
Change to assert upgrades core resolution

ericglau committed:
Add examples and docs for solidity tests

ericglau committed:
Update changelog

ericglau committed:
Update changelog dates

ericglau committed:
Fix lint

ericglau committed:
cleanup

ericglau commented on packages/core/src/cli/validate/build-info-file.ts:
```diff
@@ -172,75 +189,88 @@ async function loadBuildInfo(buildInfoFilePath: string): Promise<{
-          const outputJson = await readJSON(outputFilePath);
-          outputData = outputJson.output ?? outputJson;
-        } catch (error) {
-          throw new ValidateCommandError(
-            `Build info file ${buildInfoFilePath} does not contain output, and output file ${outputFilePath} could not be read.`,
-          );
-        }
-
-        if (outputData === undefined) {
-          throw new ValidateCommandError(
-            `Build info file ${buildInfoFilePath} does not contain output, and output file ${outputFilePath} is missing Solidity compiler output.`,
-          );
-        }
+        throw new ValidateCommandError(
+          `Build info file ${buildInfoFilePath} does not contain output, and output file ${outputFilePath} is missing Solidity compiler output.`,
+          () => HH3_BUILD_INFO_HELP,
+        );
+      }
```
not an expected case and this adds unnecessary complexity.  regardless, error would still occur below

ericglau commented on packages/core/src/cli/validate/build-info-file.ts:
```diff
@@ -172,75 +189,88 @@ async function loadBuildInfo(buildInfoFilePath: string): Promise<{
+      }
+    }
+
+    const userSourceNameMap: Record<string, string> | undefined = buildInfoJson.userSourceNameMap;
+    if (userSourceNameMap !== undefined) {
+      const canonicalToUser: Record<string, string> = {};
+      for (const [userSource, canonicalSource] of Object.entries(userSourceNameMap)) {
+        canonicalToUser[canonicalSource] = userSource;
```
Unnecessary at this point because it adds stricter malformed-input handling without evidence of an active user-facing bug

ericglau commented on packages/plugin-hardhat/examples/BoxTransparent/scripts/2-upgrade.ts:
```diff
@@ -0,0 +1,26 @@
+    process.exit(1);
+  }
+
+  const connection = await hre.network.connect();
+  const { ethers } = connection;
+  const upgradesApi = await upgrades(hre, connection);
+
+  const BoxV2 = await ethers.getContractFactory('BoxV2');
+  const box = await upgradesApi.upgradeProxy(PROXY_ADDRESS, BoxV2);
+  await box.waitForDeployment();
+
+  console.log('Box upgraded at:', await box.getAddress());
+}
```
closing does not seem necessary since these are short lived scripts/tests

ericglau commented on packages/plugin-hardhat/examples/BoxUUPS/package.json:
```diff
@@ -0,0 +1,21 @@
+    "@openzeppelin/contracts-upgradeable": "^5.0.0",
+    "@openzeppelin/hardhat-upgrades": "@openzeppelin/hardhat-upgrades@4.0.0-alpha.0",
+    "ava": "^6.0.0",
+    "ethers": "^6.8.1",
+    "hardhat": "^3.0.0",
```
this is just a minimum

ericglau commented on packages/plugin-hardhat/src/verify-proxy.ts:
```diff
@@ -60,62 +65,56 @@ interface EtherscanEventResponse {
+  hre: HardhatRuntimeEnvironment,
+  runSuper: (taskArguments: Record<string, unknown>) => Promise<unknown>,
+): Promise<unknown> {
+  const connection = await hre.network.connect();
+  const { provider } = connection;
```
We can omit for now.  It doesn't have a functional difference here because verify-proxy runs as a short-lived task against a real RPC network.

ericglau committed:
fix outside diff range issues

ericglau committed:
Update assert
